front/lib/api/files/client_executable.ts

```diff
@@ -276,3 +288,389 @@ export async function getClientExecutableFileContent(
+  );
+}
+
+function isEditFileAction(
+  action: AgentMCPActionModel
+): action is AgentMCPActionModel & {
+  augmentedInputs: { old_string: string; new_string: string };
+} {
+  return (
+    action.toolConfiguration.originalName ===
+      EDIT_CONTENT_CREATION_FILE_TOOL_NAME &&
+    typeof action.augmentedInputs.old_string === "string" &&
+    typeof action.augmentedInputs.new_string === "string"
+  );
+}
+
+function isCreateFileResourceOutput(
+  output: AgentMCPActionOutputItem
+): output is AgentMCPActionOutputItem & {
+  content: { type: "resource"; resource: { fileId: string } };
+} {
+  return (
+    output.content.type === "resource" &&
+    output.content.resource &&
+    typeof output.content.resource.fileId === "string"
+  );
+}
+
+function isRevertFileActionOutput(
+  output: AgentMCPActionOutputItem
+): output is AgentMCPActionOutputItem & {
+  content: { type: "text"; text: string };
+} {
+  return (
+    output.content.type === "text" && typeof output.content.text === "string"
+  );
+}
+
+async function fetchEditOrRevertActionsForFile(
+  auth: Authenticator,
+  fileId: string,
+  conversationId: number
+): Promise<AgentMCPActionModel[]> {
+  const workspaceId = auth.getNonNullableWorkspace().id;
+
+  return AgentMCPActionModel.findAll({
+    include: [
+      {
+        model: AgentMessage,
+        as: "agentMessage",
+        required: true,
+        include: [
+          {
+            model: Message,
+            as: "message",
+            required: true,
+            where: {
+              conversationId,
+              workspaceId,
+            },
+          },
+        ],
+      },
+    ],
+    where: {
+      workspaceId,
+      status: "succeeded",
+      [Op.and]: [
+        {
+          [Op.or]: [
+            {
+              "toolConfiguration.originalName":
+                EDIT_CONTENT_CREATION_FILE_TOOL_NAME,
+            },
+            { "toolConfiguration.originalName": REVERT_LAST_EDIT_TOOL_NAME },
+          ],
+        },
+        { "augmentedInputs.file_id": fileId },
+      ],
+    },
+    order: [["createdAt", "ASC"]],
+  });
+}
+
+async function fetchCreateActionsForConversation(
+  auth: Authenticator,
+  conversationId: number
+): Promise<AgentMCPActionModel[]> {
+  const workspaceId = auth.getNonNullableWorkspace().id;
+
+  return AgentMCPActionModel.findAll({
+    include: [
+      {
+        model: AgentMessage,
+        as: "agentMessage",
+        required: true,
+        include: [
+          {
+            model: Message,
+            as: "message",
+            required: true,
+            where: {
+              conversationId,
+              workspaceId,
+            },
+          },
+        ],
+      },
+    ],
+    where: {
+      workspaceId,
+      status: "succeeded",
+      "toolConfiguration.originalName": CREATE_CONTENT_CREATION_FILE_TOOL_NAME,
+    },
+    order: [["createdAt", "ASC"]],
+  });
+}
+
+async function getFileActions(
```

Comment:
I will try to have reproducible steps tomorrow but something is wrong in this logic, sometimes I cannot revert even though the last action was not revert. I think it happens when you revert => ask to revert further => get denied => and then edit and asking for revert